### PR TITLE
Fix: Session state stuck in programming session

### PR DESCRIPTION
**Pull request by lorinciubeica10**
lookup_state_transition_by_diagcomm_for_active lacked the default-state fallback that prevented session updates from non-default sessions.

## Checklist

- [x] I have tested my changes locally
- [ ] I have added or updated documentation
- [x] I have linked related issues or discussions
- [x] I have added or updated tests

## Related

https://github.com/eclipse-opensovd/classic-diagnostic-adapter/issues/493

### CODE REVIEW

File: cda-interfaces/src/ecumanager.rs [hunk 1]
```diff
@@ -621,13 +621,14 @@ pub trait EcuStateManager: Send + Sync + 'static {
...
     /// Lookup the transition between the active session and the requested one.
+    /// Returns the diag comm used for the transition and the canonical target session name.
     /// # Errors
     /// * `DiagServiceError::AccessDenied` if no transition exists
     /// * `DiagServiceError::NotFound` on various lookup errors.
     fn lookup_session_change(
         &self,
         session: &str,
-    ) -> impl Future<Output = Result<DiagComm, DiagServiceError>> + Send;
+    ) -> impl Future<Output = Result<(DiagComm, String), DiagServiceError>> + Send;
```
Comment (alexmohr) [inline cda-interfaces/src/ecumanager.rs]: suggestion: Please do not use string as error, revert this and keep using DiagServiceError

**Review by alexmohr — CHANGES_REQUESTED** [review]
Thanks for the change. In general looks good. One finding inline. 
Aside from this, it would be great to add an integration test or at least a unit test to prevent regressions.

**Comment by alexmohr** [comment]
```
    sovd::runtimefiles::runtimefiles_execution_responses_follow_operation_standard
```

This is a flaky test. I have fixed in #450, so you can ignore that.

File: cda-comm-uds/src/session.rs [hunk 2]
```diff
@@ -118,6 +138,7 @@ impl<S: EcuGateway, T: EcuManager> UdsSession for UdsManager<S, T> {
         let result = self
             .send_with_optional_timeout(ecu_name, dc, security_plugin, None, true, None)
             .await?;
+        update_session_state_on_positive_response(ecu_diag_service, session, &result).await;
         match result.response_type() {
```
Comment (alexmohr) [inline cda-comm-uds/src/session.rs:121]: nitpick: Why did you add `update_session_state_on_positive_response` which checks the response, when the match block below has a branch for positive response anyway? It would be more efficient to do it in the match block before the start task, so we do not have to check the response type twice in a row.

File: cda-comm-uds/src/session.rs [hunk 3]
```diff
@@ -167,3 +175,67 @@ impl<S: EcuGateway, T: EcuManager> UdsSession for UdsManager<S, T> {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (theswiftfox) [inline cda-comm-uds/src/session.rs:172]: q: i don't fully understand the reason for this tests. they do not test the check you added but rather duplicate the code block you added above? if someone changed the code the tests would still be green as they do not check the actual logic.
You would need to test `set_ecu_session` with mocked positive / negative responses to get a valuable result from the tests.

**Comment by alexmohr** [comment]
@lorinciubeica10 After you addressed the @theswiftfox comment, you can rebase onto main to fix the duplicate code check :)

File: cda-comm-uds/src/transport.rs [hunk 4]
```diff
@@ -75,7 +75,7 @@ impl<S: EcuGateway, T: EcuManager> UdsManager<S, T> {
                     variant_state = ?status.variant_state,
                     "Triggering variant detection before send"
                 );
-                if let Err(e) = self.detect_variant_if_needed(ecu_name).await {
+                if let Err(e) = self.detect_variant(ecu_name).await {
```
Comment (alexmohr) [inline cda-comm-uds/src/transport.rs]: question: Why are you changing this from `detect_variant_if_needed` to `detect_variant`, which forces a variant detection for all `send_with_optional_timeout` calls? `send_with_optional_timeout` is called in a bunch of places, not related to session handling.

Comment (lorinciubeica10) [reply cda-comm-uds/src/transport.rs]: the change from detect_variant_if_needed to detect_variant was wrong, forcing variant detection would affect unrelated requests. 
I restored detect_variant_if_needed

File: cda-comm-uds/src/session.rs [hunk 5]
```diff
@@ -167,3 +175,141 @@ impl<S: EcuGateway, T: EcuManager> UdsSession for UdsManager<S, T> {
...
+        assert_eq!(current_session(&manager).await.as_deref(), Some("Default"));
+    }
+
+    /// Regression test for the ECU appearing stuck in the programming session.
+    /// The state chart lookup that populates `ServicePayload::new_session` only
+    /// resolves transitions whose source is the active state, so once the ECU
+    /// had left the default session that fallback stopped firing and the stored
+    /// state was never overwritten again.
+    ///
+    /// The double cannot reproduce that lookup (it has no state chart), so what
```
Comment (alexmohr) [inline cda-comm-uds/src/session.rs]: q: ` The double cannot reproduce that lookup ` what do you mean by 'the double`?

Comment (lorinciubeica10) [reply cda-comm-uds/src/session.rs]: I meant test double. 
TestEcuDb is rudimentary, has no state chart and therefore cannot reproduce that lookup. 
For clarity, I can replace double with TestEcuDb.

Comment (alexmohr) [reply cda-comm-uds/src/session.rs]: > For clarity, I can replace double with TestEcuDb.

That would be great, thank you.

Comment (lorinciubeica10) [reply cda-comm-uds/src/session.rs]: fixed it

File: cda-comm-uds/src/transport.rs [hunk 6]
```diff
@@ -916,7 +916,7 @@ mod tests {
 }
 
 #[cfg(test)]
-mod send_tests {
+pub(crate) mod send_tests {
```
Comment (alexmohr) [inline cda-comm-uds/src/transport.rs:919]: suggestion: as this is re-used in another test now imho it would make sense to make the re-used functions part of a test_helpers module instead, similar to test_helpers in cda-comm-uds.

File: cda-comm-uds/src/session.rs [hunk 7]
```diff
@@ -120,6 +120,14 @@ impl<S: EcuGateway, T: EcuManager> UdsSession for UdsManager<S, T> {
             .await?;
         match result.response_type() {
             DiagServiceResponseType::Positive => {
+                ecu_diag_service
+                    .read()
+                    .await
+                    .set_service_state(
+                        cda_interfaces::service_ids::SESSION_CONTROL,
+                        session.to_owned(),
```
Comment (alexmohr) [inline cda-comm-uds/src/session.rs]: todo: Don't call `set_service_state` here. This won't apply the changes when `send_genericservice` is used. 
Also this is already tracked through `lookup_state_transition_by_diagcomm_for_active`. 
But speaking of that, this function:

`lookup_state_transition_by_diagcomm_for_active` is currently *not* doing the same default session fallback as `lookup_state_transition_for_active` so we should close this gap. Although strictly speaking it's an issue on main and I only found it due to your PR. I attached a patch, you can apply but don't have to. If not I will crate a follow up. 
In any case. These added lines do not belong here.
See cda-comm-uds/src/transport.rs:562-565 for reference:
```rust
if let Some(new_session) = payload.new_session {
    ecu_read
        .set_service_state(service_ids::SESSION_CONTROL, new_session)
        .await;
}
```

[session.patch](https://github.com/user-attachments/files/31959621/session.patch)

**Review by alexmohr — APPROVED** [review]
thank you, lgtm.
